libs/x-content/src/main/java/org/elasticsearch/xcontent/internal/XContentProviderLocator.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.xcontent.internal;
+
+import org.elasticsearch.core.internal.provider.ProviderLocator;
+import org.elasticsearch.xcontent.spi.XContentProvider;
+
+import java.util.ServiceConfigurationError;
+
+/**
+ * A provider locator for finding the {@link XContentProvider}.
+ */
+public final class XContentProviderLocator {
+
+    static final String PROVIDER_NAME = "x-content";
+
+    static final String PROVIDER_MODULE_NAME = "org.elasticsearch.xcontent.impl";
+
+    /**
+     * Returns the provider instance.
+     */
+    public static final XContentProvider INSTANCE = provider();
+
+    private static XContentProvider provider() {
+        Module m = XContentProviderLocator.class.getModule();
+        if (m.isNamed() && m.getDescriptor().uses().stream().anyMatch(XContentProvider.class.getName()::equals) == false) {
```

Comment:
Can this be moved into the locator so we don't have to duplicate it? Shouldn't the module of the locator and the provider be the same, so we can just use `providerType.getModule()` there?

Reply:
Yeah, good suggestion. I'll drop it from this PR, but address it in the modules branch.